### PR TITLE
[Win] Add Events queue

### PR DESCRIPTION
**Pull request by Mon1Amour4**

### CODE REVIEW

File: Engine/Fuego/EventQueue.h [hunk 1]
```diff
@@ -12,9 +12,10 @@ class FUEGO_API EventQueue
     FUEGO_NON_COPYABLE_NON_MOVABLE(EventQueue)
 
     virtual void Update() = 0;
-    virtual std::shared_ptr<const Event> Front() = 0;
```
Comment (LtSnail) [inline Engine/Fuego/EventQueue.h:15]: It breaks API and breaks MacOS. We should use shared_ptr here

File: Engine/Fuego/Windows/WindowWin.cpp [hunk 2]
```diff
@@ -1,26 +1,54 @@
...
+        case WM_KEYUP:
+            m_EventQueue.PushEvent(
+                std::make_unique<KeyReleasedEvent>(static_cast<int>(wparam)));
+            break;
+
         case WM_CLOSE:
-            DestroyWindow(hWnd);
+            m_EventQueue.PushEvent(std::make_unique<WindowCloseEvent>());
             break;
-        case WM_DESTROY:
```
Comment (LtSnail) [inline Engine/Fuego/Windows/WindowWin.cpp:12]: We should think how to respond to this event

File: Engine/Fuego/Windows/WindowWin.cpp [hunk 3]
```diff
@@ -1,26 +1,54 @@
-#include "WindowWin.h"
+﻿#include "WindowWin.h"
 
 namespace Fuego
 {
-LRESULT CALLBACK WindowProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+std::unordered_map<HWND, WindowWin*> WindowWin::hwndMap;
+
+LRESULT CALLBACK WindowWin::WindowProcStatic(HWND hWnd, UINT uMsg,
```
Comment (LtSnail) [inline Engine/Fuego/Windows/WindowWin.cpp:7]: Why do we need this change?

**Review by LtSnail — CHANGES_REQUESTED** [review]
Please, be careful when you change cross-platform code

File: Engine/Fuego/Windows/WindowWin.h [hunk 4]
```diff
@@ -25,9 +27,16 @@ class WindowWin : public Window
...
+    LRESULT WindowProc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam);
+    static std::unordered_map<HWND, WindowWin*> hwndMap;
+
    private:
     void Init(const WindowProps& props, EventQueue& eventQueue);
     void Shutdown();
+    static LRESULT CALLBACK WindowProcStatic(HWND hWnd, UINT uMsg,
+                                             WPARAM wParam, LPARAM lParam);
+
+    EventQueue& m_EventQueue;
```
Comment (LtSnail) [inline Engine/Fuego/Windows/WindowWin.h:39]: Why do we need a reference here?

Comment (Mon1Amour4) [reply Engine/Fuego/Windows/WindowWin.h:39]: Cause WindowProc func has to be static and at the same time I need access to WindowQueue members, so I map Static func to window api  and then I cann member func

File: Engine/Fuego/Windows/WindowWin.cpp (hunk 2 above)
Comment (Mon1Amour4) [reply Engine/Fuego/Windows/WindowWin.cpp:12]: We will, not in this commit

File: Engine/Fuego/EventQueue.h (hunk 1 above)
Comment (Mon1Amour4) [reply Engine/Fuego/EventQueue.h:15]: u meant const Event?

Comment (Mon1Amour4) [reply Engine/Fuego/EventQueue.h:15]: fixed

**Review by LtSnail — APPROVED** [review]

**Pull request merged**